File: common/src/main/java/io/github/xiewuzhiying/vs_addition/compats/create/behaviour/Link/DualLinkBehaviour.java
package io.github.xiewuzhiying.vs_addition.compats.create.behaviour.Link;

import com.simibubi.create.Create;
import com.simibubi.create.content.equipment.clipboard.ClipboardCloneable;
import com.simibubi.create.content.redstone.link.IRedstoneLinkable;
import com.simibubi.create.content.redstone.link.RedstoneLinkNetworkHandler;
import com.simibubi.create.foundation.blockEntity.SmartBlockEntity;
import com.simibubi.create.foundation.blockEntity.behaviour.BehaviourType;
import com.simibubi.create.foundation.blockEntity.behaviour.BlockEntityBehaviour;
import com.simibubi.create.foundation.blockEntity.behaviour.ValueBoxTransform;
import com.simibubi.create.foundation.utility.Couple;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.Vec3;
import org.apache.commons.lang3.tuple.Pair;

import java.util.function.Function;
import java.util.function.IntConsumer;
import java.util.function.IntSupplier;

public class DualLinkBehaviour extends BlockEntityBehaviour implements IRedstoneLinkable, ClipboardCloneable {
    public static final BehaviourType<DualLinkBehaviour> TYPE = new BehaviourType<>();

    enum Mode {
        TRANSMIT, RECEIVE
    }

    RedstoneLinkNetworkHandler.Frequency frequencyFirst;
    RedstoneLinkNetworkHandler.Frequency frequencyLast;
    public ValueBoxTransform firstSlot;
    public ValueBoxTransform secondSlot;
    Vec3 textShift;

    public boolean newPosition;
    private Mode mode;
    private IntSupplier transmission;
    private IntConsumer signalCallback;

    protected DualLinkBehaviour(SmartBlockEntity be, Pair<ValueBoxTransform, ValueBoxTransform> slots) {
        super(be);
        frequencyFirst = RedstoneLinkNetworkHandler.Frequency.EMPTY;
        frequencyLast = RedstoneLinkNetworkHandler.Frequency.EMPTY;
        firstSlot = slots.getLeft();
        secondSlot = slots.getRight();
        textShift = Vec3.ZERO;
        newPosition = true;
    }

    public static DualLinkBehaviour receiver(SmartBlockEntity be, Pair<ValueBoxTransform, ValueBoxTransform> slots,
                                             IntConsumer signalCallback) {
        DualLinkBehaviour behaviour = new DualLinkBehaviour(be, slots);
        behaviour.signalCallback = signalCallback;
        behaviour.mode = Mode.RECEIVE;
        return behaviour;
    }

    public static DualLinkBehaviour transmitter(SmartBlockEntity be, Pair<ValueBoxTransform, ValueBoxTransform> slots,
                                                IntSupplier transmission) {
        DualLinkBehaviour behaviour = new DualLinkBehaviour(be, slots);
        behaviour.transmission = transmission;
        behaviour.mode = Mode.TRANSMIT;
        return behaviour;
    }

    public DualLinkBehaviour moveText(Vec3 shift) {
        textShift = shift;
        return this;
    }

    public void copyItemsFrom(DualLinkBehaviour behaviour) {
        if (behaviour == null)
            return;
        frequencyFirst = behaviour.frequencyFirst;
        frequencyLast = behaviour.frequencyLast;
    }

    @Override
    public boolean isListening() {
        return mode == Mode.RECEIVE;
    }

    @Override
    public int getTransmittedStrength() {
        return mode == Mode.TRANSMIT ? transmission.getAsInt() : 0;
    }

    @Override
    public void setReceivedStrength(int networkPower) {
        if (!newPosition)
            return;
        signalCallback.accept(networkPower);
    }

    public void notifySignalChange() {
        Create.REDSTONE_LINK_NETWORK_HANDLER.updateNetworkOf(getWorld(), this);
    }

    @Override
    public void initialize() {
        super.initialize();
        if (getWorld().isClientSide)
            return;
        getHandler().addToNetwork(getWorld(), this);
        newPosition = true;
    }

    @Override
    public Couple<RedstoneLinkNetworkHandler.Frequency> getNetworkKey() {
        return Couple.create(frequencyFirst, frequencyLast);
    }

    @Override
    public void unload() {
        super.unload();
        if (getWorld().isClientSide)
            return;
        getHandler().removeFromNetwork(getWorld(), this);
    }

    @Override
    public boolean isSafeNBT() {
        return true;
    }

    @Override
    public void write(CompoundTag nbt, boolean clientPacket) {
        super.write(nbt, clientPacket);
        nbt.put("DualFrequencyFirst", frequencyFirst.getStack()
                .save(new CompoundTag()));
        nbt.put("DualFrequencyLast", frequencyLast.getStack()
                .save(new CompoundTag()));
        nbt.putLong("DualLastKnownPosition", blockEntity.getBlockPos()
                .asLong());
    }

    @Override
    public void read(CompoundTag nbt, boolean clientPacket) {
        long positionInTag = blockEntity.getBlockPos()
                .asLong();
        long positionKey = nbt.getLong("DualLastKnownPosition");
        newPosition = positionInTag != positionKey;

        super.read(nbt, clientPacket);
        frequencyFirst = RedstoneLinkNetworkHandler.Frequency.of(ItemStack.of(nbt.getCompound("DualFrequencyFirst")));
        frequencyLast = RedstoneLinkNetworkHandler.Frequency.of(ItemStack.of(nbt.getCompound("DualFrequencyLast")));
    }

    public void setFrequency(boolean first, ItemStack stack) {
        stack = stack.copy();
        stack.setCount(1);
        ItemStack toCompare = first ? frequencyFirst.getStack() : frequencyLast.getStack();
        boolean changed = !ItemStack.isSameItemSameTags(stack, toCompare);

        if (changed)
            getHandler().removeFromNetwork(getWorld(), this);

        if (first)
            frequencyFirst = RedstoneLinkNetworkHandler.Frequency.of(stack);
        else
            frequencyLast = RedstoneLinkNetworkHandler.Frequency.of(stack);

        if (!changed)
            return;

        blockEntity.sendData();
        getHandler().addToNetwork(getWorld(), this);
    }

    @Override
    public BehaviourType<?> getType() {
        return TYPE;
    }

    private RedstoneLinkNetworkHandler getHandler() {
        return Create.REDSTONE_LINK_NETWORK_HANDLER;
    }

    public static class SlotPositioning {
        Function<BlockState, Pair<Vec3, Vec3>> offsets;
        Function<BlockState, Vec3> rotation;
        float scale;

        public SlotPositioning(Function<BlockState, Pair<Vec3, Vec3>> offsetsForState,
                               Function<BlockState, Vec3> rotationForState) {
            offsets = offsetsForState;
            rotation = rotationForState;
            scale = 1;
        }

        public SlotPositioning scale(float scale) {
            this.scale = scale;
            return this;
        }

    }

    public boolean testHit(Boolean first, Vec3 hit) {
        BlockState state = blockEntity.getBlockState();
        Vec3 localHit = hit.subtract(Vec3.atLowerCornerOf(blockEntity.getBlockPos()));
        return (first ? firstSlot : secondSlot).testHit(state, localHit);
    }

    @Override
    public boolean isAlive() {
        Level level = getWorld();
        BlockPos pos = getPos();
        if (blockEntity.isChunkUnloaded())
            return false;
        if (blockEntity.isRemoved())
            return false;
        if (!level.isLoaded(pos))
            return false;
        return level.getBlockEntity(pos) == blockEntity;
    }

    @Override
    public BlockPos getLocation() {
        return getPos();
    }

    @Override
    public String getClipboardKey() {
        return "Frequencies";
    }

    @Override
    public boolean writeToClipboard(CompoundTag tag, Direction side) {
        tag.put("First", frequencyFirst.getStack()
                .save(new CompoundTag()));
        tag.put("Last", frequencyLast.getStack()
                .save(new CompoundTag()));
        return true;
    }

    @Override
    public boolean readFromClipboard(CompoundTag tag, Player player, Direction side, boolean simulate) {
        if (!tag.contains("First") || !tag.contains("Last"))
            return false;
        if (simulate)
            return true;
        setFrequency(true, ItemStack.of(tag.getCompound("First")));
        setFrequency(false, ItemStack.of(tag.getCompound("Last")));
        return true;
    }

}

File: common/src/main/kotlin/io/github/xiewuzhiying/vs_addition/compats/create/behaviour/link/DualLinkBehaviour.kt
package io.github.xiewuzhiying.vs_addition.compats.create.behaviour.link

import com.simibubi.create.Create
import com.simibubi.create.content.equipment.clipboard.ClipboardCloneable
import com.simibubi.create.content.redstone.link.IRedstoneLinkable
import com.simibubi.create.content.redstone.link.RedstoneLinkNetworkHandler
import com.simibubi.create.foundation.blockEntity.SmartBlockEntity
import com.simibubi.create.foundation.blockEntity.behaviour.BehaviourType
import com.simibubi.create.foundation.blockEntity.behaviour.BlockEntityBehaviour
import com.simibubi.create.foundation.blockEntity.behaviour.ValueBoxTransform
import com.simibubi.create.foundation.utility.Couple
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.nbt.CompoundTag
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.phys.Vec3
import org.apache.commons.lang3.tuple.Pair
import org.valkyrienskies.core.api.ships.Ship
import org.valkyrienskies.mod.common.getShipManagingPos
import org.valkyrienskies.mod.common.util.toJOML
import org.valkyrienskies.mod.common.util.toMinecraft
import java.util.function.Function
import java.util.function.IntConsumer
import java.util.function.IntSupplier


open class DualLinkBehaviour protected constructor(
    be: SmartBlockEntity?,
    slots: Pair<ValueBoxTransform?, ValueBoxTransform?>
) :
    BlockEntityBehaviour(be), IRedstoneLinkable,
    ClipboardCloneable {
    internal enum class Mode {
        TRANSMIT, RECEIVE
    }

    var frequencyFirst: RedstoneLinkNetworkHandler.Frequency
    var frequencyLast: RedstoneLinkNetworkHandler.Frequency
    var firstSlot: ValueBoxTransform
    var secondSlot: ValueBoxTransform
    var textShift: Vec3

    var newPosition: Boolean = true
    private var mode: Mode? = null
    private var transmission: IntSupplier? = null
    private var signalCallback: IntConsumer? = null

    init {
        frequencyFirst = RedstoneLinkNetworkHandler.Frequency.EMPTY
        frequencyLast = RedstoneLinkNetworkHandler.Frequency.EMPTY
        firstSlot = slots.left!!
        secondSlot = slots.right!!
        textShift = Vec3.ZERO
    }

    fun moveText(shift: Vec3): DualLinkBehaviour {
        textShift = shift
        return this
    }

    fun copyItemsFrom(behaviour: DualLinkBehaviour?) {
        if (behaviour == null) return
        frequencyFirst = behaviour.frequencyFirst
        frequencyLast = behaviour.frequencyLast
    }

    override fun isListening(): Boolean {
        return mode == Mode.RECEIVE
    }

    override fun getTransmittedStrength(): Int {
        return if (mode == Mode.TRANSMIT) transmission!!.asInt else 0
    }

    override fun setReceivedStrength(networkPower: Int) {
        if (!newPosition) return
        signalCallback!!.accept(networkPower)
    }

    fun notifySignalChange() {
        Create.REDSTONE_LINK_NETWORK_HANDLER.updateNetworkOf(world, this)
    }

    override fun initialize() {
        super.initialize()
        if (world.isClientSide) return
        handler.addToNetwork(world, this)
        newPosition = true
    }

    override fun getNetworkKey(): Couple<RedstoneLinkNetworkHandler.Frequency> {
        return Couple.create(frequencyFirst, frequencyLast)
    }

    override fun unload() {
        super.unload()
        if (world.isClientSide) return
        handler.removeFromNetwork(world, this)
    }

    override fun isSafeNBT(): Boolean {
        return true
    }

    override fun write(nbt: CompoundTag, clientPacket: Boolean) {
        super.write(nbt, clientPacket)
        nbt.put(
            "DualFrequencyFirst", frequencyFirst.stack
                .save(CompoundTag())
        )
        nbt.put(
            "DualFrequencyLast", frequencyLast.stack
                .save(CompoundTag())
        )
        nbt.putLong(
            "DualLastKnownPosition", blockEntity.blockPos
                .asLong()
        )
    }

    override fun read(nbt: CompoundTag, clientPacket: Boolean) {
        val positionInTag = blockEntity.blockPos
            .asLong()
        val positionKey = nbt.getLong("DualLastKnownPosition")
        newPosition = positionInTag != positionKey

        super.read(nbt, clientPacket)
        frequencyFirst = RedstoneLinkNetworkHandler.Frequency.of(ItemStack.of(nbt.getCompound("DualFrequencyFirst")))
        frequencyLast = RedstoneLinkNetworkHandler.Frequency.of(ItemStack.of(nbt.getCompound("DualFrequencyLast")))
    }

    fun setFrequency(first: Boolean, stack: ItemStack) {
        val stack2 = stack.copy()
        stack2.count = 1
        val toCompare = if (first) frequencyFirst.stack else frequencyLast.stack
        val changed = !ItemStack.isSameItemSameTags(stack2, toCompare)

        if (changed) handler.removeFromNetwork(world, this)

        if (first) frequencyFirst = RedstoneLinkNetworkHandler.Frequency.of(stack2)
        else frequencyLast = RedstoneLinkNetworkHandler.Frequency.of(stack2)

        if (!changed) return

        blockEntity.sendData()
        handler.addToNetwork(world, this)
    }

    override fun getType(): BehaviourType<*> {
        return TYPE
    }

    private val handler: RedstoneLinkNetworkHandler
        get() = Create.REDSTONE_LINK_NETWORK_HANDLER

    class SlotPositioning(
        var offsets: Function<BlockState, org.apache.commons.lang3.tuple.Pair<Vec3, Vec3>>,
        var rotation: Function<BlockState, Vec3>
    ) {
        var scale: Float = 1f

        fun scale(scale: Float): SlotPositioning {
            this.scale = scale
            return this
        }
    }

    fun testHit(first: Boolean, hit: Vec3): Boolean {
        val state = blockEntity.blockState

        val level: Level? = this.world
        var pos1 = hit
        var pos2 = Vec3.atLowerCornerOf(blockEntity.blockPos)

        if (level != null) {
            val ship1: Ship? = level.getShipManagingPos(pos1)
            val ship2: Ship? = level.getShipManagingPos(pos2)
            if (ship1 != null && ship2 == null) {
                pos2 = ship1.worldToShip.transformPosition(pos2.toJOML()).toMinecraft()
            } else if (ship1 == null && ship2 != null) {
                pos1 = ship2.worldToShip.transformPosition(pos1.toJOML()).toMinecraft()
            }
        }

        val localHit = pos1.subtract(pos2)
        return (if (first) firstSlot else secondSlot).testHit(state, localHit)
    }

    override fun isAlive(): Boolean {
        val level: Level = world
        val pos = pos
        if (blockEntity.isChunkUnloaded) return false
        if (blockEntity.isRemoved) return false
        if (!level.isLoaded(pos)) return false
        return level.getBlockEntity(pos) === blockEntity
    }

    override fun getLocation(): BlockPos {
        return pos
    }

    override fun getClipboardKey(): String {
        return "Frequencies"
    }

    override fun writeToClipboard(tag: CompoundTag, side: Direction?): Boolean {
        tag.put(
            "First", frequencyFirst.stack
                .save(CompoundTag())
        )
        tag.put(
            "Last", frequencyLast.stack
                .save(CompoundTag())
        )
        return true
    }

    override fun readFromClipboard(tag: CompoundTag, player: Player?, side: Direction?, simulate: Boolean): Boolean {
        if (!tag.contains("First") || !tag.contains("Last")) return false
        if (simulate) return true
        setFrequency(true, ItemStack.of(tag.getCompound("First")))
        setFrequency(false, ItemStack.of(tag.getCompound("Last")))
        return true
    }

    companion object {
        @JvmStatic
        val TYPE: BehaviourType<DualLinkBehaviour> = BehaviourType()

        @JvmStatic
        fun receiver(
            be: SmartBlockEntity?, slots: Pair<ValueBoxTransform?, ValueBoxTransform?>,
            signalCallback: IntConsumer?
        ): DualLinkBehaviour {
            val behaviour = DualLinkBehaviour(be, slots)
            behaviour.signalCallback = signalCallback
            behaviour.mode = Mode.RECEIVE
            return behaviour
        }

        @JvmStatic
        fun transmitter(
            be: SmartBlockEntity?, slots: Pair<ValueBoxTransform?, ValueBoxTransform?>,
            transmission: IntSupplier?
        ): DualLinkBehaviour {
            val behaviour = DualLinkBehaviour(be, slots)
            behaviour.transmission = transmission
            behaviour.mode = Mode.TRANSMIT
            return behaviour
        }
    }
}
